### ISSUE TITLE
Optimise SSR under heavy load

### ISSUE THREAD

**Issue by tobi-or-not-tobi** [SAP/spartacus#8755]
SSR is causing issues in a few occasions: 
-  out of memory –  in case a pages cannot be rendered soon enough, SSR takes too much memory and eventually fails.
- failover – when there's no response in time, a 500 error is returned. 

These symptoms happen mainly when the system is under heavy load. This is often due to an unhealthy implementation (memory leaks) or when the SSR response is not cached.

While mitigations for these scenario's can be implemented in ccv2, there are a few reasons to do this in the (Spartacus) SSR implementation: 
- it provides a flexible solution that can be adjusted over time
- it provides a customisable solution, that can be further enhanced by customers
- ability to support the solution directly from Spartacus
- agnostic to alternative hosts (non ccv2)
- not being blocked by ccv2 roadmap, architecture, priorities

## Solution design
SSR will be optimised to serve pages instantly at any time. 

- SSR will queue incoming requests.
- Only the first queued page will be rendered.
- The first queued page request will be served with the rendered page if it is available within a certain time (configuration required). If time is passed, the CSR app (`index.html`) is served instead. The time will be configurable, so that customers can set it 0 to always instantly response with a page. The default will however be x-seconds (2 I'd say), to ensure that local devs will not be confused with CSR response from SSR.
- Sub-sequential requests for the same page will be served with the CSR app (`index.html`) instantly. 
CSR app (`index.html`) response is served with the `Cache-Control:no-store` header, as the cache layer should never store this response for sub-sequential requests. 
- The rendered page can be stored in memory for sub-sequential requests, but by default it won't. Storing the page in memory could introduce massive garbage collection. We expect the system to either be capable to serve the page instantly within timeout (local dev server), or have a cache layer on top of SSR (production system). When cache is in place (i.e. CDN), sub-sequential requests for the same page will not hit SSR, as the cache layer will return them from memory. 

## Additional optimisations (out of scope) 
In production, a few additional items should be considered
- Warmup script, that hits all pages of the storefront. This will warm up SSR and the cache. 
- (configurable) cache control headers in SSR that will expose the expiration of the SSR pages. This should be configurable per page, as the underlying content of the page will dictate the expiration time. 
- SSR could interact with the API through internal URLs rather than be routed externally. This is not straightforward as the API is often load balanced, and the load balancer is only accessible through external routing.
- Add logging information so that monitoring tools can provide operation data to the response.

**Comment by giancorderoortiz** [comment]
It's now possible to actually measure any improvements done to SSR. 
Dynatrace cluster is now available for testing purposes. 
Demo server has 2.0 CSR and SSR deployed and has dynatrace's oneagent installed. 
QA server or dev3, I believe, could be used for comparison purposes. 

For generating load, we can use gatling or cypress e2e tests for homepage. 
In the case of gatling (i.e. scala), we can run hundreds of simultaneous users accessing home page and observe the failure rate. 
In the case of cypress, we could run the same homepage tests in sequence, but this wouldn't be exactly load testing. 

@tobi-or-not-tobi, @dunqan Please let me know how to proceed. 

Load testing with CCV2 is a bit more complex and involves financial considerations. 
I would suggest testing with Linux dev-servers first.

**Comment by giancorderoortiz** [comment]
How many concurrent users is considered heavy load? 
What are your immediate expectations after improvements in this ticket are implemented?
By this I mean, what meaningful metric should we establish so we can actually measure an improvement under 'heavy load' ?

**Comment by giancorderoortiz** [comment]
Compare 2.0 against improvements made in this ticket. 
Trace through using Dynatrace.
